src/server/service/kv.rs

```diff
@@ -318,6 +318,47 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
         ctx.spawn(future);
     }
 
+    fn kv_refresh_lock(
+        &mut self,
+        ctx: RpcContext<'_>,
+        mut req: RefreshLockRequest,
+        sink: UnarySink<RefreshLockResponse>,
+    ) {
+        let timer = GRPC_MSG_HISTOGRAM_VEC.kv_refresh_lock.start_coarse_timer();
+
+        let (cb, f) = paired_future_callback();
+        let mut options = Options::default();
+        options.lock_ttl = req.get_ttl();
+        options.txn_size = req.get_txn_size();
+        let res = self.storage.async_refresh_lock(
+            req.take_context(),
+            req.take_key(),
+            req.get_start_version(),
+            options,
+            cb,
+        );
+        let future = AndThenWith::new(res, f.map_err(Error::from))
+            .and_then(move |v| {
+                let mut resp = RefreshLockResponse::new();
+                if let Some(err) = extract_region_error(&v) {
+                    resp.set_region_error(err);
+                } else {
+                    match v {
+                        Ok(()) => resp.set_ttl(req.get_ttl()),
```

Comment:
Return the new TTL?

Reply:
I think it makes more sense to set the ttl to that returned by `MvccTxn::refresh_lock`. Maybe it's better to add a new StorageCb..

Reply:
good idea.